bugtool/cmd/bugtool.go

```
@@ -0,0 +1,187 @@
+	log.Infof("Running scheduled tasks")
+	if err := root.Run(ctx, runtime); err != nil {
+		log.WithError(err).Fatal("Failed to run dump")
+	}
+	log.Debug("Draining scheduler")
+	ts, err := sched.Drain()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to schedule dump tasks")
+	}
+	log.Infof("Done, collecting results")
+	failed := 0
+	for _, t := range ts {
+		if t.Err() != nil {
+			failed++
+			log.Debugf("Scheduled task %q returned an error: %v", t.String(), t.Err())
+		}
+	}
+	if failed > 0 {
+		log.Infof("%d scheduled tasks failed", failed)
+	}
+
+	runtime.GetResults()
+
+	createArchive(config, bugtool.outDir)
+	fmt.Printf(disclaimer)
+}
+
+func generateDirName(config *options.Config) string {
+	// Prevent collision with other directories
+	nowStr := time.Now().Format("20060102-150405.999-0700-MST")
+	var prefix string
+	if config.ArchivePrefix != "" {
+		prefix = fmt.Sprintf("%s-cilium-bugtool-%s-", config.ArchivePrefix, nowStr)
+	} else {
+		prefix = fmt.Sprintf("cilium-bugtool-%s-", nowStr)
+	}
+	return prefix
+}
+
+func createArchive(config *options.Config, dbgDir string) {
+	sendArchiveToStdout := config.DumpPath == "-"
+	log.Debugf("Attempting to archive of type %q at path %q",
+		config.ArchiveType, config.DumpPath)
+	if config.Archive {
+		switch config.ArchiveType {
+		case "gz":
+			gzipPath, err := CreateGzip(dbgDir, sendArchiveToStdout)
+			if err != nil {
+				log.Fatalf("Failed to create gzip: %v", err)
+			}
+			// Note: Cilium CLI sysdump depends on this specific formatting
+			// by using a regexp to capture the output filename.
+			//
+			// TODO: Deprecate and remove
+			fmt.Printf("GZIP at %s\n", gzipPath)
+		case "tar":
+			archivePath, err := CreateArchive(dbgDir, sendArchiveToStdout)
+			if err != nil {
+				log.Fatalf("Failed to create tar: %v", err)
+			}
+			fmt.Printf("TAR at %s\n", archivePath)
+		}
+	} else {
+		fmt.Fprintf(os.Stderr, "\nDIRECTORY at %s\n", dbgDir)
+	}
+}
+
+func waitForAgentReady(ctx context.Context) {
```

this function is awkward: if it times out, it just exits. It probably should return an error instead.